src/aind_slims_api/operations/waterlog.py

```diff
@@ -0,0 +1,201 @@
+"""Common operations for logging mouse water and weight with the SLIMS API."""
+
+from datetime import datetime
+from typing import Optional
+import os
+import logging
+
+from pydantic import BaseModel, model_validator, validate_call
+
+from aind_slims_api import models, SlimsClient, exceptions
+from aind_slims_api.models import (
+    SlimsMouseContent,
+    SlimsWaterlogResult,
+    SlimsWaterRestrictionEvent,
+    SlimsUser,
+)
+from aind_slims_api.models.base import SlimsBaseModel
+
+logger = logging.getLogger(__name__)
+
+
+class Mouse:
+    """Class for tracking mouse water/weight data and syncing with SLIMS
+
+    Examples
+    --------
+
+    >>> from aind_slims_api import SlimsClient
+
+    >>> client = SlimsClient()
+    >>> user = client.fetch_model(SlimsUser,username='username')
+    >>> barcode = '000000'
+
+    >>> mouse = Mouse(barcode,user,client)
+
+    >>> print(f'Mouse {barcode} baseline weight is {mouse.details.baseline_weight_g} g')
+
+    >>> if mouse.details.water_restricted:
+    ...     mouse.switch_to_adlib_water()
+
+    >>> mouse.add_waterlog_record(
+    ...     weight=20,
+    ...     water_earned=0,
+    ...     water_supplement_recommended=1,
+    ...     water_supplement_delivered=1,
+    ... )
+
+    >>> mouse.post_baseline_weight(20)
+
+    >>> mouse.switch_to_water_restricted()
+
+    >>> mouse.update_target_weight_fraction(0.9)
+
+    """
+
+    client: SlimsClient
+    user: SlimsUser
+    barcode: str
+
+    details: SlimsMouseContent
+    waterlog_results: list[SlimsWaterlogResult]
+    restriction: SlimsWaterRestrictionEvent
+    all_restrictions: list[SlimsWaterRestrictionEvent]
+
+    def __init__(self, barcode: str, user: SlimsUser, slims_client=None):
+        """Fetch data from Slims for mouse with barcode={mouse_name}"""
+        self.client = slims_client or SlimsClient()
+        self.barcode = barcode
+        self.user = user
+
+        self.details: SlimsMouseContent = None
+        self.waterlog_results: list[SlimsWaterlogResult] = []
+        self.restriction: SlimsWaterRestrictionEvent = None
+        self.all_restrictions: list[SlimsWaterRestrictionEvent] = []
+
+        self.link_mouse: str = None
+        self.link_restrictions: str = None
+        self.link_wl_records: str = None
+
+        self._fetch_data()
+        self._fetch_pks()
+
+    def _fetch_data(self):
```

Comment:
Maybe we can implement these as properties and lazily fetch them. I can anticipate this class being used to get more than just water-related information; behavior sessions, surgical procedures, etc. Maybe it's not good to do all these fetches on instance init.